src/entities/tradeMining.ts

```diff
@@ -0,0 +1,155 @@
+import { providers, Contract, BigNumber as BN } from 'ethers';
+import { PendleTradeMiningQuerySet } from '@pendle/subgraph-sdk';
+import { NetworkInfo } from '../networks';
+import { distributeConstantsByNetwork, getCurrentTimestamp } from '../helpers';
+import { TokenAmount } from './tokenAmount';
+import { Token } from './token';
+
+export type Valuation = {
+  currency: String; // 'USD'
+  amount: String | Number; // '10.43'
+};
+
+export type TradeMiningUser = {
+  wallet: String; // wallet address
+  tradedVolume: Valuation;
+};
+
+type DefaultTradeMiningQUery = {
+  house: string;
+  phase: string;
+};
+
+type TradeMiningQuery = {
+  house: string;
+  phase: string;
+  numberOfTraders: number;
+};
+
+type GetUserRankQuery = {
+  house: string;
+  phase: string;
+  walletAddress: string;
+};
+
+type GetUserRankResponse = {
+  rank: number;
+  tradedVolume: Valuation;
+};
+
+export class TradeMining {
+  public methods(
+    signer: providers.JsonRpcSigner,
+    chainId?: number
+  ): Record<string, any> {
+    const networkInfo: NetworkInfo = distributeConstantsByNetwork(chainId);
+    const pendleSubgraphSdk = new PendleTradeMiningQuerySet(chainId || 43114);
+
+    const getTopTraders = async ({
+      phase,
+      house,
+      numberOfTraders = 10,
+    }: TradeMiningQuery): Promise<TradeMiningUser[]> => {
+      const topTraders = await pendleSubgraphSdk.getTopTraders(
+        phase,
+        house,
+        numberOfTraders
+      );
+
+      return topTraders.tradeMiningUsers.map((res: any) => {
+        return {
+          wallet: res.userAddress,
+          tradedVolume: {
+            currency: 'USD',
+            amount: parseFloat(res.volumeUSD).toFixed(2),
+          },
+        };
+      });
+    };
+
+    const getUserRank = async (
+      query: GetUserRankQuery
+    ): Promise<GetUserRankResponse> => {
+      const topTraders = await getTopTraders({
+        house: query.house,
+        phase: query.phase,
+        numberOfTraders: 1000,
+      });
+      let user = await pendleSubgraphSdk.getUserRank(
+        query.phase,
+        query.house,
+        query.walletAddress.toLowerCase()
+      );
+
+      if (user.tradeMiningUsers.length === 0)
+        return {
+          rank: -1,
+          tradedVolume: {
+            currency: 'USD',
+            amount: '0',
+          },
+        };
+
+      user = user.tradeMiningUsers[0];
+
+      let rank = 0;
+      for (const trader of topTraders) {
```

Comment:
Consider using Javascript's Array.find(element, index) instead of for loop

Reply:
Updated to use Array.find